compiler/rustc_mir_transform/src/liveness.rs

```diff
@@ -72,6 +72,14 @@ pub(crate) fn check_liveness<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> Den
         return DenseBitSet::new_empty(0);
     }
 
+    // Don't run unused pass for items generated by foreign macros
+    if tcx.def_span(parent).from_expansion()
+        && let Some(macro_def_id) = tcx.def_span(parent).ctxt().outer_expn_data().macro_def_id
+        && !macro_def_id.is_local()
```

Comment:
Why `parent` and not `def_id`?
Should this use `tcx.def_span(parent).in_external_macro(tcx.sess.source_map())`?

Reply:
use `def_id` will remove the warning for this testcase:
https://github.com/chenyukang/rust/blob/f943606503b142b24edb2b20f77f1eb737fb9ed2/tests/ui/proc-macro/attributes-included.rs#L17
yeah, `tcx.def_span(parent).in_external_macro(tcx.sess.source_map())` is simpler.